compiler/rustc_builtin_macros/src/deriving/enums.rs

```diff
@@ -0,0 +1,341 @@
+    Mutability, NestedMetaItem, PatKind, Path, PathSegment, Stmt, StmtKind, Ty, Unsafe, Variant,
+    VariantData, Visibility, VisibilityKind,
+};
+use rustc_attr::{find_repr_attrs, int_type_of_word};
+use rustc_data_structures::thin_vec::ThinVec;
+use rustc_errors::struct_span_err;
+use rustc_expand::base::{Annotatable, ExtCtxt};
+use rustc_span::symbol::{sym, Ident};
+use rustc_span::{Span, Symbol, DUMMY_SP};
+
+pub fn expand_from_repr(
+    cx: &mut ExtCtxt<'_>,
+    derive_span: Span,
+    _mitem: &MetaItem,
+    item: &Annotatable,
+    push: &mut dyn FnMut(Annotatable),
+) {
+    if let Some(ctx) = Ctx::extract(cx, item, derive_span) {
+        push(make_from_repr(cx, ctx.enum_def, ctx.ty, ctx.repr_ty));
+    }
+}
+
+struct Ctx<'enumdef> {
+    enum_def: &'enumdef EnumDef,
+    ty: P<Ty>,
+    repr_ty: P<Ty>,
+}
+
+impl<'enumdef> Ctx<'enumdef> {
+    fn extract(
+        cx: &mut ExtCtxt<'_>,
+        item: &'enumdef Annotatable,
+        derive_span: Span,
+    ) -> Option<Ctx<'enumdef>> {
+        match *item {
+            Annotatable::Item(ref annitem) => match annitem.kind {
+                ItemKind::Enum(ref enum_def, _) => {
+                    let repr = extract_repr(cx, annitem, enum_def, derive_span)?;
+
+                    let ty = cx.ty_ident(DUMMY_SP, annitem.ident);
+                    let repr_ty = cx.ty_ident(DUMMY_SP, Ident { name: repr, span: DUMMY_SP });
+
+                    return Some(Self { enum_def, ty, repr_ty });
+                }
+                _ => {}
+            },
+            _ => {}
+        }
+        struct_span_err!(
+            &cx.sess.parse_sess.span_diagnostic,
+            derive_span,
+            FIXME,
+            "`FromRepr` can only be derived for enums",
+        )
+        .span_label(item.span(), "Not an enum")
+        .emit();
+        None
+    }
+}
+
+fn extract_repr(
+    cx: &mut ExtCtxt<'_>,
+    annitem: &P<Item>,
+    def: &EnumDef,
+    derive_span: Span,
+) -> Option<Symbol> {
+    let reprs: Vec<_> = annitem
+        .attrs
+        .iter()
+        .flat_map(|attr| find_repr_attrs(&cx.sess, attr))
+        .filter_map(|r| {
+            use rustc_attr::*;
+            match r {
+                ReprInt(rustc_attr::IntType::UnsignedInt(int_type)) => Some(int_type.name()),
+                ReprInt(rustc_attr::IntType::SignedInt(int_type)) => Some(int_type.name()),
+                ReprC | ReprPacked(..) | ReprSimd | ReprTransparent | ReprAlign(..)
+                | ReprNoNiche => None,
+            }
+        })
+        .collect();
+    if reprs.len() != 1 {
+        let mut err = struct_span_err!(
+            &cx.sess.parse_sess.span_diagnostic,
+            derive_span,
+            FIXME,
+            "`FromRepr` can only be derived for enums with exactly one explicit integer representation",
+        );
+        if reprs.is_empty() {
+            err.span_label(annitem.span, "Enum missing repr");
```

Comment:
I like the rewording, but I'm 50:50 on the suggestion - while I think we _can_ give a suggestion, I'm not sure we _should_ - this PR effectively makes the repr type a part of the public API of the enum, so I'm inclined to put a placeholder into the suggestion (to prompt them to think about what the repr _should_ be looking forwards), rather than guessing based on what fits there now... WDYT?

Reply:
Works for me. My rule of thumb is that for anything that we can obviously figure out what the user intended, we should provide a suggestion, and for cases where more nuance and thought is needed, we don't, hopefully with some extra text and even a link to try and expose them to the potential consequences.